rfcs/0000-c-unwind-abi.md

```diff
@@ -126,6 +133,28 @@ may be "sandwiched" between Rust frames, so that Rust `panic`s may safely
 unwind the C++ frames, if the Rust code declares both the C++ entrypoint and
 the Rust entrypoint using `"C unwind"`.
 
+## "Plain Old Frames"
+[POF-definition]: #plain-old-frames
+
+A "POF", or "Plain Old Frame", is defined as a frame that can be trivially
+deallocated: returning from or unwinding a POF cannot cause any
+observable effects. This means that POFs do not contain any pending destructors
+(live `Drop` objects) or `catch_unwind` calls.
+
+The terminology is intentionally akin to [C++'s "Plain Old Data"
+types][cpp-POD-definition], which are types that, among other requirements, are
+trivially destructible (their destructors do not cause any observable effects,
+and may be elided as an optimization).
+
+Rust frames that do contain pending destructors or `catch_unwind` calls are
+called non-POFs.
+
+Note that a non-POF may _become_ a POF, for instance if all `Drop` objects are
+moved out of scope, or if its only `catch_unwind` call is in a code path that
+will not be executed.
```

Comment:
I think it'd be worth giving an example here, and I have to admit I still feel sort of hesitant. I wonder if we can call in an LLVM expert to validate this. Let me give a specific example of the situation here:

```rust
fn foo(c: bool) {
    let v = vec![1, 2, 3];
    if c {
        drop(v);
    }
    func_call_that_longjmps(c);
}

/// Only does a longjmp if `c` is true.
extern "C" fn func_call_that_longjmps(c: bool);
```

In this situation, the call to `func_call_that_longjmps` will (on windows, anyway) unwind with a forced exception, despite being marked as "nounwind" and despite the fact that the compile will generate a landing pad (that ultimately has no effect when executing). I believe our claim is that this is not going to be UB.

Actually, I realize as I write this that I haven't really *read* the LLVM definition -- @Amanieu maybe you can provide a citation to the relevant webdocs? I also realize that, from LLVM's POV, the existence of a landing pad is quite likely irrelevant.

So I am starting to be convinced, but I think it would be good for us to include some of this reasoning in the RFC itself.

Reply:
From the [LangRef](https://llvm.org/docs/LangRef.html):
> `nounwind`
>    This function attribute indicates that the function never raises an exception. If the function does raise an exception, its runtime behavior is undefined. However, functions marked nounwind may still trap or generate asynchronous exceptions. Exception handling schemes that are recognized by LLVM to handle asynchronous exceptions, such as SEH, will still provide their implementation defined semantics.

This is a bit vague but basically implies that `longjmp` out of a `nounwind` function has "implementation defined semantics". I have no idea what that is actually supposed to mean though...

Reply:
OK, so I guess there are two considerations:

* On Windows, where we actually *do* generate an exception, what happens. I agree that is vague but suggestive that this is "defined behavior".
* On Linux, where there is no exception, what happens. I believe the idea *here* is that LLVM (unlike Rust) doesn't itself consider it UB to "pop" a frame, even if there are pending cleanups or whatever (e.g., if you called a "C unwind" function which invoked longjmp, that could happen).

Reply:
@nikomatsakis What would be the next step here? Do you still want to find someone with more LLVM expertise (or authority) to confirm that "popping" frames with landing pads is not inherently UB?

Reply:
@nikomatsakis I've added an example based on your code above, and some explanatory verbiage, including a paragraph about the `nounwind` optimization. Please let me know if what I've written is not what you were looking for!

Reply:
I'm afraid I know no more than you about LLVM semantics. Might be best to bring this up with the LLVM people? They [have a Discourse](https://llvm.discourse.group/) these days. :D

Reply:
Yes, that seems good.